build-setup/linux.md

```diff
@@ -245,58 +245,103 @@ CMake version 3.5 is the minimum required to build PyTorch. Download version 3.1
 chmod +x cmake-3.19.3-Linux-x86_64.sh
 sudo mkdir /usr/local/cmake
 sudo ./cmake-3.19.3-Linux-x86_64.sh --skip-license --prefix=/usr/local/cmake
-```  
+```
 
-Please ensure `/usr/local/cmake/bin` is in your `PATH` environment variable. 
+Please ensure `/usr/local/cmake/bin` is in your `PATH` environment variable.
 
-### Python 
+### Python
 
 PyTorch currently requires Python 3.6, 3.7 or 3.8, and version 3.7 appears to cause fewest problems in their test status matrix, so we use that. If your system does not have a requisite version of Python install it with a package manager or build the last 3.7 release from source by downloading `Python-3.7.9.tgz` from <https://www.python.org/ftp/python/3.7.9/Python-3.7.9.tgz> then extract and build:
 
 ```
 tar -xzf Python-3.7.9.tgz
-cd Python-3.7.9/
-./configure --enable-optimizations
+cd Python-3.7.9
+./configure --prefix=/usr/local/gcc93 --enable-optimizations
 make
 sudo make altinstall
 ```
 
-### PyTorch 1.7.1
+### PyTorch 1.8.0
 
 PyTorch requires that certain Python modules are installed. Install these modules with `pip` using the same Python version you will build PyTorch with. If you followed the instructions above and built Python from source use `python3.7`:
 
 ```
-sudo /usr/local/bin/python3.7 -m pip install install numpy ninja pyyaml setuptools cffi typing_extensions future six requests dataclasses
+sudo /usr/local/gcc93/bin/python3.7 -m pip install install numpy ninja pyyaml setuptools cffi typing_extensions future six requests dataclasses
+```
+
+For aarch64 the `ninja` module is not available, so use:
+
+```
+sudo /usr/local/gcc93/bin/python3.7 -m pip install install numpy pyyaml setuptools cffi typing_extensions future six requests dataclasses
 ```
 
 Then obtain the PyTorch code:
+
 ```
-git clone --depth=1 --branch=v1.7.1 git@github.com:pytorch/pytorch.git
+git clone --depth=1 --branch=v1.8.0 git@github.com:pytorch/pytorch.git
 cd pytorch
 git submodule sync
 git submodule update --init --recursive
 ```
 
+Edit `torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp` and replace all
+occurrences of `system(` with `strlen(`. This file is used to compile
+fused CPU kernels, which we do not expect to be doing and never want to
+do for security reasons. Replacing the calls to `system()` ensures that
+a heuristic virus scanner looking for potentially dangerous function
+calls in our shipped product will not encounter these functions that run
+external processes.
+
+If you are building on aarch64, also edit `aten/src/ATen/cpu/vml.h` and
+change lines 88-93 from:
+
+```
+    parallel_for(0, size, 2048, [out, in](int64_t begin, int64_t end) {           \
```

Comment:
The macro you changed is exactly the same in v.1.7.1 so I’m guessing something else has changed that  both triggers the bug and means openmp is required. We are not doing training and evaluation is single threaded (as far as I know in our uses) so we shouldn’t really need openmp. 

Is it because of the tangled dependencies in the build system? Did you try `USE_OPENMP=OFF` or `WITH_OPENMP=OFF`?

Reply:
> evaluation is single threaded (as far as I know in our uses)

Oh, I didn't know that.

So it seems that the difference between my build of 1.8.0 and your build of 1.7.1 is that mine decided to enable OPENMP by default whereas yours didn't.  That would explain both this problem and also that fact that `libgomp` was getting linked.

If we really don't want any multi-threading in the library we ship I can rebuild it all again with `USE_OPENMP=OFF`.  That should then mean that we don't need to hack the code while using gcc 9.3 and don't need to ship `libgomp`.

So I guess it depends how certain we are that multithreading isn't giving us any benefits at all in the way we're using the library...

What do you think?

Reply:
According to https://pytorch.org/docs/stable/notes/cpu_threading_torchscript_inference.html model interference _can_ be multi-threaded.  So I think we are better off sticking with a version of the library built with `USE_OPENMP=ON` (which is the default).